File: 3657-check-if-grid-can-be-cut-into-sections/check-if-grid-can-be-cut-into-sections.cpp
class Solution {
public:
    bool countCuts(vector<pair<int, int>>& coordinates) {
        int cnt = 0, prevEnd = 0;
        for (auto& [start, end] : coordinates) {
            if (start >= prevEnd) {
                cnt++;
                if (cnt > 2) {
                    return true;
                }
            }
            prevEnd = max(prevEnd, end);
        }
        return false;
    }

    bool checkValidCuts(int n, vector<vector<int>>& rectangles) {
        int r = rectangles.size();
        vector<pair<int, int>> xCoordinates(r), yCoordinates(r);
        for (int i = 0; i < r; i++) {
            xCoordinates[i] = {rectangles[i][0], rectangles[i][2]};
            yCoordinates[i] = {rectangles[i][1], rectangles[i][3]};
        }
        
        sort(xCoordinates.begin(), xCoordinates.end());
        if (countCuts(xCoordinates)) {
            return true;
        }
        
        sort(yCoordinates.begin(), yCoordinates.end());
        if (countCuts(yCoordinates)) {
            return true;
        }
        
        return false;
    }
};

File: 3657-check-if-grid-can-be-cut-into-sections/check-if-grid-can-be-cut-into-sections.py
class Solution:
    def checkValidCuts(self, n: int, rectangles: List[List[int]]) -> bool:
        def count_cuts(coordinates):
            cnt = prev_end = 0
            for start, end in coordinates:
                if start >= prev_end:
                    cnt += 1
                prev_end = max(prev_end, end)
            return cnt

        r = len(rectangles)
        x_coordinates = [(rectangles[i][0], rectangles[i][2]) for i in range(r)]
        x_coordinates.sort()
        if count_cuts(x_coordinates) >= 3:
            return True
        y_coordinates = [(rectangles[i][1], rectangles[i][3]) for i in range(r)]
        y_coordinates.sort()
        if count_cuts(y_coordinates) >= 3:
            return True
        return False
